File: sol/solution/0301-0400/0301/Solution.scala
object Solution {
    def removeInvalidParentheses(s: String): List[String] = {
        var rl = 0
        var rr = 0
        for (c <- s) {
            if (c == '(') rl += 1
            else if (c == ')') {
                if (rl > 0) rl -= 1
                else rr += 1
            }
        }
        val ans = scala.collection.mutable.Set[String]()
        def dfs(i: Int, l: Int, r: Int, open: Int, path: StringBuilder): Unit = {
            if (i == s.length) {
                if (l == 0 && r == 0 && open == 0) ans += path.toString
                return
            }
            val c = s(i)
            val len = path.length
            if (c == '(') {
                if (l > 0) dfs(i + 1, l - 1, r, open, path)
                path.append(c)
                dfs(i + 1, l, r, open + 1, path)
                path.setLength(len)
            } else if (c == ')') {
                if (r > 0) dfs(i + 1, l, r - 1, open, path)
                if (open > 0) {
                    path.append(c)
                    dfs(i + 1, l, r, open - 1, path)
                    path.setLength(len)
                }
            } else {
                path.append(c)
                dfs(i + 1, l, r, open, path)
                path.setLength(len)
            }
        }
        dfs(0, rl, rr, 0, new StringBuilder)
        ans.toList
    }
}

File: sol/solution/0301-0400/0301/Solution.kt
class Solution {
    fun removeInvalidParentheses(s: String): List<String> {
        val res = mutableListOf<String>()
        dfs(s, 0, 0, res, false)
        return res
    }
    private fun dfs(
        s: String,
        deletionSearch: Int,
        stackSearch: Int,
        res: MutableList<String>,
        ri: Boolean
    ) {
        var stack = 0
        var p = stackSearch
        while (p < s.length && stack >= 0) {
            when (s[p]) {
                '(' -> stack++
                ')' -> stack--
            }
            p++
        }
        if (stack < 0) {
            val prefix = s.substring(0, p)
            var i = deletionSearch
            while (i < prefix.length) {
                if (s[i] == ')' && (i == deletionSearch || s[i - 1] != ')')) {
                    dfs(s.removeRange(i, i + 1), i, p - 1, res, ri)
                }
                i++
            }
        } else {
            if (!ri) {
                dfs(reverseInvert(s), 0, 0, res, true)
            } else {
                res.add(reverseInvert(s))
            }
        }
    }
    private fun reverseInvert(s: String): String {
        val sb = StringBuilder()
        for (c in s) {
            when (c) {
                '(' -> sb.append(')')
                ')' -> sb.append('(')
                else -> sb.append(c)
            }
        }
        return sb.reverse().toString()
    }
}